### PR TITLE
feat: validate CLI installation and version during activation of CLI Integration extension

### PR DESCRIPTION
**Pull request by daphne-sfdc**
### What does this PR do?
This PR includes new functionality to check whether a supported CLI version is installed on the user's computer.  There are 4 main cases that are being caught: (1) CLI not installed (2) Outdated version of SFDX CLI that is incompatible with our extensions (3) Only SF v1 is installed (4) Both SFDX v7 and SF v2 are installed.  In each of those 4 cases, the CLI Integration extension will not activate, and the user will not be able to see SFDX commands in the command palette.  The user will receive an error notification about the unsupported/missing CLI when the activation of the CLI Integration extension fails.

### What issues does this PR fix or reference?
@W-14196710@

### Functionality Before
As a result of the recent switch from SFDX -> SF-style variables to get rid of warnings in the Output tab, users with a CLI version below SFDX v7.193.2 are no longer able to use the extensions because older CLI versions do not support SF-style variables.  The extensions will activate, but the older version is caught at runtime.
Screenshot 2023-10-24 at 7 00 13 PM

### Functionality After
If the user does not have CLI installed on their computer or their installed CLI version is no longer supported, they will get an error notification when they first create/open a Salesforce Project and the CLI Integration extension will not activate.  Based on the case, one of these 4 error messages will pop up:
Screenshot 2023-11-28 at 9 37 21 AM

### CODE REVIEW

File: packages/salesforcedx-vscode-core/src/messages/i18n.ja.ts
Comment (daphne-sfdc) [inline packages/salesforcedx-vscode-core/src/messages/i18n.ja.ts:252]: I updated the English error message in i18n.ts but the Japanese error message is still the old one. @sbudhirajadoc

File: packages/salesforcedx-vscode-core/src/messages/i18n.ja.ts
Comment (daphne-sfdc) [inline packages/salesforcedx-vscode-core/src/messages/i18n.ja.ts:254]: This is a new error message.  The English version is in i18n.ts but I don't have a Japanese translation so I temporarily copied it over and added 3 exclamation points in front. @sbudhirajadoc

File: packages/salesforcedx-vscode-core/src/messages/i18n.ts
Comment (daphne-sfdc) [inline packages/salesforcedx-vscode-core/src/messages/i18n.ts:293]: I changed the error message because SF v1 does not map sf commands to sfdx and thus cannot be used by our extensions.  The isCliInstalled() function in cliConfiguration.ts uses "which sfdx" for the checking, which means having only SF v1 is caught at the same time as having no CLI installed.

File: packages/salesforcedx-vscode-core/src/messages/i18n.ts
Comment (CristiCanizales) [inline packages/salesforcedx-vscode-core/src/messages/i18n.ts:1]: As discussed on huddle, could you make these 2 messages a bit more similar for consistency?

**Review by CristiCanizales — CHANGES_REQUESTED** [review]
LGTM :) Waiting on the tests for approval. Thanks Daphne! 🥳

File: packages/salesforcedx-vscode-core/src/messages/i18n.ts
Comment (daphne-sfdc) [reply packages/salesforcedx-vscode-core/src/messages/i18n.ts:1]: Fixed

File: packages/salesforcedx-utils/src/cli/checkCliVersion.ts
Comment (gbockus-sf) [inline packages/salesforcedx-utils/src/cli/checkCliVersion.ts:12]: unit test me

File: packages/salesforcedx-utils/src/cli/checkCliVersion.ts
Comment (gbockus-sf) [inline packages/salesforcedx-utils/src/cli/checkCliVersion.ts:64]: How about using an enum for these return values?   Can be used here, in the code using the return value, and in test

File: packages/salesforcedx-utils/src/cli/checkCliVersion.ts
Comment (gbockus-sf) [inline packages/salesforcedx-utils/src/cli/checkCliVersion.ts:50]: have you run prettier against this?  The if/else with spaces in between seems....non-ideal.

File: packages/salesforcedx-utils/src/cli/checkCliVersion.ts
Comment (gbockus-sf) [inline packages/salesforcedx-utils/src/cli/checkCliVersion.ts:41]: Let's use constants (in this file is fine) for the cli version numbers.

File: packages/salesforcedx-utils/src/cli/checkCliVersion.ts
Comment (gbockus-sf) [inline packages/salesforcedx-utils/src/cli/checkCliVersion.ts:54]: same story..constants for sf version numbers

File: packages/salesforcedx-utils/src/cli/checkCliVersion.ts
Comment (gbockus-sf) [inline packages/salesforcedx-utils/src/cli/checkCliVersion.ts:63]: Yeah I'd def go with 
`} else { // Case 3: ......`

File: packages/salesforcedx-vscode-core/src/index.ts
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-core/src/index.ts:572]: So keeping all the logic out of the activation method is a good thing (think separation of concerns).  Can we refactor this to be a stand alone utility function so all we see here is `await validateDependancies();`?  It will throw if all deps aren't correct and resolve happily otherwise.

Comment (gbockus-sf) [reply packages/salesforcedx-vscode-core/src/index.ts:572]: Also throwing empty errors is sadness (it's gets reported to telemetry).  We should throw a custom error or an error with a valid message

**Review by gbockus-sf — COMMENTED** [review]
some feedback for your consideration

File: packages/salesforcedx-utils/src/cli/checkCliVersion.ts
Comment (diyer) [reply packages/salesforcedx-utils/src/cli/checkCliVersion.ts:50]: should we check both the versions instead of an else if for the issue we are discussing here https://salesforce-internal.slack.com/archives/C041ZPJH317/p1699052602748759?thread_ts=1699051206.563149&cid=C041ZPJH317

Comment (diyer) [reply packages/salesforcedx-utils/src/cli/checkCliVersion.ts:58]: have we considered addressing having both sfdx and sf installed in this change?

Comment (daphne-sfdc) [reply packages/salesforcedx-utils/src/cli/checkCliVersion.ts:58]: @diyer I can try to make the change now and revert if @gbockus-sf thinks it is unnecessary.

Comment (diyer) [reply packages/salesforcedx-utils/src/cli/checkCliVersion.ts:58]: lets wait for @gbockus-sf suggestion Daphne. It would be nice to address that issue as well IMHO

File: packages/salesforcedx-utils/src/cli/checkCliVersion.ts
Comment (daphne-sfdc) [reply packages/salesforcedx-utils/src/cli/checkCliVersion.ts:78]: @gbockus-sf I just ran Prettier against that file and Prettier actually didn't complain about the if/else with spaces in between.  I tried switching to the formatting that you suggested in this comment but when I saved the file, Prettier moved the comment to the next line (see attached screenshot for Option 3).  What would you prefer that I do?

1. Keep the original formatting with the spaces between the if/else statements
Screenshot 2023-11-09 at 4 48 00 PM

2. Turn off Prettier for the file and reformat to what you suggested
Screenshot 2023-11-09 at 4 49 09 PM

3. Let Prettier auto-format the comments and keep them inside the brackets on the next line
Screenshot 2023-11-09 at 4 42 34 PM

Comment (CristiCanizales) [reply packages/salesforcedx-utils/src/cli/checkCliVersion.ts:78]: I like either 1 or 3 🙂

**Comment by CristiCanizales** [comment]
QE notes [here](https://salesforce.quip.com/5YyEA6PetzcG)

File: packages/salesforcedx-utils/src/cli/checkCliVersion.ts
Comment (daphne-sfdc) [inline packages/salesforcedx-utils/src/cli/checkCliVersion.ts:123]: I didn't do another check for SF versions because for SF v2 anything above 2.0.0 is supported.  The case of only SF v1 has already been caught in Case 2.

File: packages/salesforcedx-utils/src/cli/checkCliVersion.ts
Comment (daphne-sfdc) [inline packages/salesforcedx-utils/src/cli/checkCliVersion.ts:105]: This is checking that the SF version is v2 and the SFDX version is not the same as the SF version.  Because SF v2 maps sf to sfdx, the results of `sfdx --version` and `sf --version` should be identical.  If they are different, that means the user has both SFDX v7 and SF v2 installed simultaneously.

**Review by CristiCanizales — CHANGES_REQUESTED** [review]
LGTM 😄

File: packages/salesforcedx-utils/src/cli/checkCliVersion.ts
Comment (diyer) [reply packages/salesforcedx-utils/src/cli/checkCliVersion.ts:66]: @daphne-sfdc are we not taking care of both sfdx and sf installed?

File: packages/salesforcedx-vscode-core/src/index.ts
Comment (diyer) [reply packages/salesforcedx-vscode-core/src/index.ts:731]: +1

File: packages/salesforcedx-vscode-core/src/index.ts
Comment (diyer) [inline packages/salesforcedx-vscode-core/src/index.ts:730]: are these user facing errors? if yes then can they also suggest the right course of action to take ?

Comment (daphne-sfdc) [reply packages/salesforcedx-vscode-core/src/index.ts:730]: No, those are for telemetry purposes and are not user facing.  The user will see the error in the **showErrorNotification()** message, which you can find the text in **i18n.ts**.

File: packages/salesforcedx-vscode-core/src/constants.ts
Comment (diyer) [inline packages/salesforcedx-vscode-core/src/constants.ts:16]: nit: can this variable name be also updated to SF_CLI_...?

File: packages/salesforcedx-utils/src/cli/checkCliVersion.ts
Comment (diyer) [inline packages/salesforcedx-utils/src/cli/checkCliVersion.ts:110]: what is this check for sfCliVersionArray[0] < minSFVersionArray[0] ?

File: packages/salesforcedx-utils/src/cli/checkCliVersion.ts
Comment (diyer) [inline packages/salesforcedx-utils/src/cli/checkCliVersion.ts:104]: cant this be simplified to following cases:
1. sfdx and sf both not installed
2. sf v1 installed
3. if sfdx installed, check if sfdx version is supported
4. if sf v2 is installed, check sf version is supported

Comment (daphne-sfdc) [reply packages/salesforcedx-utils/src/cli/checkCliVersion.ts:104]: Resolved during our 1:1 on Monday 12/4.

Summary of the discussion: Case 3 (Both SFDX CLI (v7) and SF CLI (v2) are installed at the same time) is necessary for semantic reasons.  If the user has an outdated version of SFDX v7 that they installed simultaneously with a valid version of SF v2, the VSCode Terminal prefers the outdated version of SFDX v7, even though it is incompatible with the extensions.  If we remove Case 3, the user will see the error message that they have an outdated version of SFDX but get confused when they log a Github issue and tell us that they have SF v2 yet the extensions are not working.  This extra check for semantics is to remove that confusion by giving a clear error message to the user about what is happening with their system.

**Review by sbudhirajadoc — COMMENTED** [review]
Update based on discussions here -- https://docs.google.com/document/d/1lE-xJMhEXfcrenaTU2LYSR6HB0HnGP6P-VW7KfPZ0tI/edit

**Review by sbudhirajadoc — APPROVED** [review]

File: packages/salesforcedx-vscode-core/src/util/cliConfiguration.ts
Comment (CristiCanizales) [inline packages/salesforcedx-vscode-core/src/util/cliConfiguration.ts:36]:
```suggestion
    'salesforce_cli_not_found',
```
I think we should update this too

File: packages/salesforcedx-vscode-core/src/index.ts
Comment (CristiCanizales) [inline packages/salesforcedx-vscode-core/src/index.ts:548]: I'm not sure about this 💀 because we have so many users on windows... and we would be 100% skipping the validation for them. Is this something we want? It seems like half a feature

File: packages/salesforcedx-utils/test/jest/cli/checkCliVersion.test.ts
Comment (peternhale) [inline packages/salesforcedx-utils/test/jest/cli/checkCliVersion.test.ts:26]: This test is not necessary given class does not have a formal constructor

File: packages/salesforcedx-vscode-core/src/index.ts
Comment (peternhale) [inline packages/salesforcedx-vscode-core/src/index.ts:718]: This function does not need promises

File: packages/salesforcedx-utils/src/cli/checkCliVersion.ts
Comment (peternhale) [inline packages/salesforcedx-utils/src/cli/checkCliVersion.ts:19]: Consider renaming this class w/o a verb in the name, such as CliVersion. Class names tend to have noun based names. Same would apply to the enum.

File: packages/salesforcedx-utils/test/jest/cli/checkCliVersion.test.ts
Comment (peternhale) [inline packages/salesforcedx-utils/test/jest/cli/checkCliVersion.test.ts:1]: I don't see tests for the getSfdxCliVersion or get SfCliVersion functions.

Comment (daphne-sfdc) [reply packages/salesforcedx-utils/test/jest/cli/checkCliVersion.test.ts:1]: I originally tried to write tests for getSfdxCliVersion() and getSfCliVersion() functions but found that they're dependent on the version installed on the computer and thus I can't put definitive values to check for.  For example, suppose I have SF v2.14.1 installed on my computer.  In order to get the tests to pass, both getSfdxCliVersion() and getSfCliVersion() need to check for SF v2.14.1.  But the CLI team made a release and the next person to run the tests has SF v2.15.0.  Because v2.14.1 ≠ v2.15.0, the second person has two failing tests, even though they have a valid CLI version.  That's just one example - a third person can have SFDX v7 installed, which is another case.  It's difficult to test getSfdxCliVersion() and getSfCliVersion() functions unless we just test that _something_ is returned, regardless of the value.

Comment (diyer) [reply packages/salesforcedx-utils/test/jest/cli/checkCliVersion.test.ts:1]: @daphne-sfdc cant you use mocks?

File: packages/salesforcedx-utils/src/cli/checkCliVersion.ts
Comment (diyer) [inline packages/salesforcedx-utils/src/cli/checkCliVersion.ts:63]: do we have to run this version number by cli team? I know you ran some manual tests to make sure this version is supported, but is that a valid way to test this is indeed the minimum supported version or is there other authoritative way.

File: packages/salesforcedx-utils/test/jest/cli/checkCliVersion.test.ts
Comment (daphne-sfdc) [reply packages/salesforcedx-utils/test/jest/cli/checkCliVersion.test.ts]: @diyer I didn't think of that.  Let me check the other unit test classes to see how they use mocks and try to implement mocking in this test class.

**Review by peternhale — CHANGES_REQUESTED** [review]

File: packages/salesforcedx-utils/src/cli/checkCliVersion.ts
Comment (daphne-sfdc) [reply packages/salesforcedx-utils/src/cli/checkCliVersion.ts]: @diyer 

Some background info -
In the last several SFDX v7 versions, the CLI team has been gradually updating the CLI to allow the use of SF-style commands.  For example, the SF-style command for authorizing an org is `sfdx org login web`.  If you have a CLI version that is SFDX v7.191.1 or newer, you are able to run this command.  If you have an older version of the CLI, you will get the error message `Warning: org login web is not a sfdx command.` and you are prevented from running the command.

Our VSCode extensions need to use these new SF-style commands because the CLI team is gradually deprecating the SFDX-style commands.  For example, in the latest version SF v2.21.8, the CLI team has removed support for the SFDX-style command for authorizing an org: `sfdx force auth login web`.  If you have SF v2.21.8 and try to run that command, you get the error message `Warning: force auth login web is not a sf command.` and you are blocked from running the command.  In order to support newer CLI versions, we must keep up with the updates from SFDX -> SF style.

To answer the question -
As we make more and more SFDX -> SF style changes in our VSCode extensions, the last supported CLI version will gradually increase.  From my testing of some common commands, such as deploy/retrieve, auth, run Apex tests, etc. the current last supported version is SFDX v7.193.2.  I suspect that this minimum version will eventually not be supported because we are seeing more PRs for SFDX -> SF changes, so it would be good to reevaluate this number once every few months to make sure it shouldn't be raised, and also to recommend to customers that they make the leap to SF v2.

If we are unsure about whether SFDX v7.193.2 is the correct minimum version and there is still the possibility of seeing SF-style commands not being supported at runtime, I would say it's better to be safe than sorry and choose a higher version than the what we think is the minimum.  So for example, if we think SFDX v7.193.2 _might_ be supported but we're not 100% sure, I would put the minimum as a higher version like v7.209.6 (the latest SFDX v7 version).  That way, users with older CLI versions will all receive the message that their CLI is no longer supported, and we force them to upgrade to SF v2.  If we do this, we're guaranteed to not run into any SFDX-style command issues but some customers will have CLI versions that tell them are outdated but which really don't break the system.  It's the tradeoff of prematurely forcing customers to upgrade to SF v2 vs. possibly still seeing runtime errors and getting Github issues about them.

As this is VSCode extensions-specific, I don't think we need to run this number by CLI team - we can read the release notes for SFDX here: https://github.com/forcedotcom/cli/blob/main/releasenotes/sfdx/README.md.  From my testing, I think the minimum supported version is v7.193.2 as I put in the code, but feel free to do further testing before this PR gets merged if you think it is necessary and/or we can come back and reevaluate this minimum version in a few months.

File: packages/salesforcedx-utils/src/cli/cliVersionStatus.ts
Comment (peternhale) [inline packages/salesforcedx-utils/src/cli/cliVersionStatus.ts:22]: @daphne-sfdc this function and the next have the same logic, with diff in the command being run. See if this can be reduced to one function that manages the execSync and then use the current functions to call into the new function passing command to run.

**Review by peternhale — CHANGES_REQUESTED** [review]

File: packages/salesforcedx-vscode-core/src/messages/i18n.ts
Comment (diyer) [inline packages/salesforcedx-vscode-core/src/messages/i18n.ts:302]: As we talked in slack, lets make the tone of this message bit subtle :) https://salesforce-internal.slack.com/archives/C054SJJAB24/p1703132241482919

**Review by CristiCanizales — APPROVED** [review]
All good from QA perspective both in mac and windows! ✅ 
After the text updates I'd say :shipit: 
Nice work Daphne! 🥳 
QA doc [here](https://salesforce.quip.com/5YyEA6PetzcG)

File: packages/salesforcedx-utils/src/cli/cliVersionStatus.ts
Comment (peternhale) [inline packages/salesforcedx-utils/src/cli/cliVersionStatus.ts:34]:
``` suggestion
  public getCliVersion(isSfdx: boolean): string {
    try {
      const result = execSync(`${isSfdx ? 'sfdx' : 'sf'} --version`);
      return result.toString();
    } catch {
      return 'No CLI';
    }
  }
```

**Review by peternhale — CHANGES_REQUESTED** [review]

**Review by peternhale — APPROVED** [review]

**Comment by sbudhirajadoc** [comment]
The text for the message recommending the move to SF from SFDX is [here](https://docs.google.com/document/d/1lE-xJMhEXfcrenaTU2LYSR6HB0HnGP6P-VW7KfPZ0tI/edit).

**Pull request merged**
